Support Kitkat

consp1racy opened the pull request:
This is another attempt at suporting API 19. We're developing for a market that's often stuck 10 years ago and seeing this work felt amazing. Maybe label it as experimental.

What this consists of:
* Rebase the original `api19` branch on top of `master`.
* Polyfill `android.system.Os.write(FileDescriptor, ByteBuffer)`.
* Backport `MediaCodec.getOutputBuffer` API.

What doesn't work:
* Emulators. See https://issuetracker.google.com/issues/36982354

What definitely works:
* HTC Desire 601
* Will try and expand the list as I test on more devices.

The performance isn't amazing but it works.

![htc desire 601](https://user-images.githubusercontent.com/4589773/174561883-f9f5ea2e-3c07-490a-a7e0-e30dbc56a417.png)

rom1v commented on server/build.gradle:
```diff
@@ -21,6 +21,7 @@ android {
 dependencies {
     implementation fileTree(dir: 'libs', include: ['*.jar'])
     testImplementation 'junit:junit:4.13.1'
+    compileOnly 'androidx.annotation:annotation:1.3.0' // 1.3.0 doesn't depend on Kotlin stdlib.
```
Nope, the server should be buildable without any java/kotlin dependencies (and without gradle): [`build_without_gradle.sh`](https://github.com/Genymobile/scrcpy/blob/master/server/build_without_gradle.sh)

(Btw, I just removed the `implementation fileTree` line above: 7f2f5950f2624c1a3fd847af6248bcbef71a3124)

consp1racy replied:
Hmm, is it OK to download a prebuilt into a directory (e.g. `prebuilt-compileOnly`) and also use everything in it as classpath for `javac`? That should work both ways. I really want the IDE to help me not call API 19 stuff on API 21.

> and without gradle

Any other reason than using build system within build system is weird? (And the flex of knowing how to manually produce an APK? 😁 )

rom1v replied:
> is it OK to download a prebuilt into a directory

Nope. If it was a necessary dependency, we could discuss what to do, but here they are just annotations, it does not worth the dependency.

> Any other reason than using build system within build system is weird?

It is used to build the official Debian package: https://salsa.debian.org/yangfl-guest/scrcpy/-/blob/ab5d09e733bb9635b4410693ceff95dbc859dd4c/debian/rules#L41
There were problems with gradle/kotlin (Debian provides its own packages for android tools).

rom1v reviewed (commented):
Thank you for your work.

I guess you followed #98 discussions. After re-reading it, when I concluded:

> The video encoder does not work below 5.0.

Maybe it was related to the smart watch from the issue (without an hardware encoder) and not to Android < 5 after all. So indeed, it might be possible to support Android 4.4.

TBH, I'm not sure it's worth it to add support for 4.4 in 2022. It adds complexity and maintenance (btw there is still probably the `dalvik-cache` creation to handle). But if it works, maybe we could merge (anyway at least a separate up-to-date `api19` branch could be useful)… I have no Android 4.4 device to test unfortunately. I will try to find one.

consp1racy commented on server/src/main/java/android/system/Os.java:
```diff
@@ -0,0 +1,53 @@
+ * limitations under the License.
+ */
+
+package android.system;
```
Makes sense, let's do an `OsCompat`, as long as it's just one method and not the whole class it should be manageable. And it'll play nicely with the `@*Api` annotations. I'll either do the OO thing again or what they do in AndroidX Core. There has to be an indirection between the call to new API to avoid slow class verification.

consp1racy replied:
I will have to put `libcore.io` package in a separate `compileOnly` module and make it work without gradle too. We don't want to ship that inside the package.

consp1racy commented on server/src/main/java/com/genymobile/scrcpy/MediaCodecCompat.java:
```diff
@@ -0,0 +1,129 @@
+ * Version of {@link MediaCodec} that backports {@link #getOutputBuffer} to Kitkat.
+ * The backported implementation isn't thread safe.
+ */
+abstract class MediaCodecCompat {
```
The beautiful thing about this is that `ScreenEncoder` basically changed by one line in exchange for API 19 support. I don't want to pollute `ScreenEncoder` with compat logic.

Actually, I plan to extract a `MediaCodecListCompat` as well to keep `ScreenEncoder` clean.

consp1racy commented on server/src/main/java/com/genymobile/scrcpy/MediaCodecCompat.java:
```diff
@@ -0,0 +1,129 @@
+
+        private void validateOutputByteBuffer(
+                @Nullable ByteBuffer[] buffers, int index, @NonNull MediaCodec.BufferInfo info) {
+            if (buffers != null && index >= 0 && index < buffers.length) {
```
Come to think about it, this may be necessary only for "invalidated" buffers which had position/limit reset. The hidden API isn't even available on Kitkat. I'll remove this method entirely. Thanks!

consp1racy commented:
I've made it so the AndroidX `annotation-1.3.0.jar` lives as a copy inside the repo. It's a compile-only dependency both in `server/build.gradle` and `server/build_without_gradle.sh`. That way it contributes to developer experience in Android Studio, it contributes to lint, but has no effect on the produced artifact. **The annotations and their calls are not present `scrcpy-server` binary.**

The same thing now applies to the `libcore.io` package. This package is part of the framework and only needs to be presetn at compile time. The binary doesn't contain any of these classes.

The same approach should be done in the future with the system AIDL interfaces (e.g. `IRotationWatcher.aidl` or `IClipboard`) these are part of the framework and shouldn't be part of the binary.

`android.system` package remains shadowed for now. Tried doing it better but it bloated the binary.

Removed unnecessary logic from `MediaCodecCompat`.

Extracted `MediaCodecListCompat` from `ScreenEncoder` to keep it clean.

consp1racy commented:
* We no longer shadow `android.system` package. Instead there's new `androidx.system` package.
* `OsCompat.write(FileDescriptor, ByteBuffer)` now advances buffer position on all platforms. This was fixed in Lollopop MR1. This is no longer part of `IO.writeFully`.
* `OsCompat.write` automatically retries on `EINTR`.
* As a side-effect, we can now use lambdas.

yume-chan commented:
Why the video socket is using file descriptors, not standard Java streams? Is file descriptor faster than stream or something?

consp1racy commented:
> Why the video socket is using file descriptors, not standard Java streams? Is file descriptor faster than stream or something?

https://github.com/Genymobile/scrcpy/commit/66def38b734829170654647086bd76f99efdec70

hebing0305 commented:
I really need it. How can I download it?
